language-plutus-core/src/Language/PlutusCore/Core/Instance/Eq.hs

```diff
@@ -117,18 +113,22 @@ eqTermM (Var _ name1) (Var _ name2) =
     eqNameM name1 name2
 eqTermM (Constant _ con1) (Constant _ con2) =
     eqM con1 con2
-eqTermM (Builtin _ bi1) (Builtin _ bi2) =
-    eqM bi1 bi2
-eqTermM LamAbs{}   _ = empty
-eqTermM TyAbs{}    _ = empty
-eqTermM IWrap{}    _ = empty
-eqTermM Apply{}    _ = empty
-eqTermM Unwrap{}   _ = empty
-eqTermM Error{}    _ = empty
-eqTermM TyInst{}   _ = empty
-eqTermM Var{}      _ = empty
-eqTermM Constant{} _ = empty
-eqTermM Builtin{}  _ = empty
+eqTermM (ApplyBuiltin _ bn1 tys1 args1) (ApplyBuiltin _ bn2 tys2 args2) = do
+    eqM bn1 bn2
+    guard $ length tys1 == length tys2
+    zipWithM_ eqTypeM tys1 tys2
+    guard $ length args1 == length args2
+    zipWithM_ eqTermM args1 args2
```

Comment:
Actually, I have a feeling that having

```haskell
data OneIsLonger = LeftIsLonger | RightIsLonger

zipExact :: [a] -> [b] -> Either OneIsLonger [(a, b)]
```

we can make this code and the type checker's code nicely readable. Especially the type checke, because there we won't need some explicit recursion then.